rustdoc: hide `#[repr]` if it isn't part of the public ABI

fmease opened the pull request:
Follow-up to rust-lang/rust#115439.
Unblocks rust-lang/rust#116743, CC @dtolnay.

Fixes rust-lang/rust#66401.
Fixes rust-lang/rust#128364.
Fixes rust-lang/rust#137440.

Only display the representation `#[repr(REPR)]` (where `REPR` is not `Rust` or `transparent`) of a given type if none of its variants (incl. the synthetic variants of structs) are `#[doc(hidden)]` and all of its fields are public and not `#[doc(hidden)]` since it's likely not meant to be considered part of the public ABI otherwise.

`--document-{private,hidden}-items` works as expected in this context, too.

Moreover, we now also factor in the presence of `#[doc(hidden)]` when checking whether to show `repr(transparent)` or not.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6558887681/job/17813363398) [(plain)](https://github.com/rust-lang/rust/commit/1608e1c00737ddcfd0df7d75e1662dc90b285f1d/checks/17813363398/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=fmease
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_46e46b39-ea6d-4711-a68a-5e4f0570f5c9
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=9d3253f9708fd9c15d22783e0d4df019c0c81618
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_46e46b39-ea6d-4711-a68a-5e4f0570f5c9
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_46e46b39-ea6d-4711-a68a-5e4f0570f5c9
GITHUB_TRIGGERING_ACTOR=fmease
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116882/merge
GITHUB_WORKFLOW_SHA=9d3253f9708fd9c15d22783e0d4df019c0c81618
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
.......... (100/109)
........  (109/109)


/checkout/tests/rustdoc-gui/item-decl-colors.goml item-decl-colors... FAILED
[ERROR] (line 26 from line 40) ".item-decl .code-attribute" not found: for command `assert-css: (".item-decl .code-attribute", {"color": |attr_color|}, ALL)`
[ERROR] (line 26 from line 54) ".item-decl .code-attribute" not found: for command `assert-css: (".item-decl .code-attribute", {"color": |attr_color|}, ALL)`
[ERROR] (line 26 from line 68) ".item-decl .code-attribute" not found: for command `assert-css: (".item-decl .code-attribute", {"color": |attr_color|}, ALL)`
FError: ()
Build completed unsuccessfully in 0:02:19
  local time: Wed Oct 18 09:51:08 UTC 2023
  network time: Wed, 18 Oct 2023 09:51:08 GMT
```

</details>

fmease commented on src/librustdoc/clean/types.rs:
```diff
@@ -743,67 +741,111 @@ impl Item {
         }
-        attrs
+
+        let is_public = cache.document_private
```
I should return early here if it's the default repr / `repr(Rust)` for perf reasons.

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -743,67 +741,122 @@ impl Item {
+            || if adt.is_enum() {
+                // FIXME(fmease): Should we take the visibility of fields of variants into account?
+                // FIXME(fmease): `any` or `all`?
+                adt.variants().is_empty()
```
I think I'd go for `any` in both cases.

fmease replied:
@RalfJung, does

```js
public($REPR)
    <=>
ANY_OR_EMPTY V: Variants($ADT).
    ANY_OR_EMPTY F: Fields(V).
        public(F)
```

sound good to you as well? Or should it be `ANY_OR_EMPTY ALL` (current impl) or `ALL ALL` or `ALL ANY_OR_EMPTY`.

RalfJung replied:
Usually for structs, if there is at least one private field then we say you can't rely on the struct staying how it is. For instance if your `repr(transparent)` relies on another type being a ZST and that type has at least one private field, we warn about that (and we eventually want to make that an error).

So I'd say the same should go for the repr. If any field is private, then the repr is private.

RalfJung replied:
There's no such things as private enum variants (unfortunately).

fmease replied:
Oops, I meant `doc(hidden)` variants ^^'

RalfJung replied:
I'm not sure which rules people usually use for when an enum is not-semver-stable. `#[non_exhaustive]` is one thing. (A `#[non_exhaustive]`  struct should probably also have its `repr` hidden.) All fields of all variants are always public so that's probably the only thing one can do?

RalfJung replied:
> Oops, I meant doc(hidden) variants ^^'

Ah. that's an attribute I don't have intuition for, sorry. In rustc it doesn't affect anything.

Can one add this to a field? So on a struct with a doc(hidden) field, should the repr be shown? No idea.

dtolnay replied:
> on a struct with a doc(hidden) field, should the repr be shown?

Here is an example of what goes wrong if you show repr on a struct with doc(hidden) field.

```rust
#[repr(C)]
pub struct Struct {
    pub first: u8,
    #[doc(hidden)]
    pub second: u16,
    pub third: u8,
}
```

<img width="506" alt="Screenshot 2023-10-18 at 3 41 44 PM" src="https://github.com/rust-lang/rust/assets/1940490/34b439ab-5d62-4cad-9703-e0a8e8f01fa2">

Rustdoc purports a repr(C) struct in which the first byte is `first`, the second byte is `third`, and some other fields follow. Given the Rust-like syntax in which rustdoc shows #\[repr(C)\], this feels misleading. For a struct that is _actually_ this:

```rust
#[repr(C)]
pub struct Struct {
    pub first: u8,
    pub third: u8,
    // ... other fields ...
}
```

one would expect they can cast `&Struct` to `&[u8; 2]` and read `first` and `third` from it. If they do that in this case though, they get UB from looking at a padding byte.

I think this would be a useful bar to keep in mind as a minimum desirable property; rustdoc should not show a repr in such a way that misleads reader about reality. That does not necessarily need to mean hiding such reprs, though that might be the most expedient path forward. Alternatively rustdoc could be more discerning about placing the /\*private field\*/ comment in between the correct fields when there is a repr.

fmease replied:
Hmm, do we already track this in a GitHub issue? With the (unstable) introduction of `std::mem::offset_of`, it feels like we should up the priority of it. If I remember correctly, it'd need quite a bit of rewiring inside rustdoc to render `/* private field */` in the correct order for `repr(C)` structs since those fields are *stripped* early.

fmease replied:
RalfJung, re taking `doc(hidden)` on variants into account when computing the visibility of a `repr`, I've included that to hide the `repr(u8)` on `core::ffi::c_void` which has two `doc(hidden)` enum variants.

RalfJung replied:
Sounds to me then like we'd want to hide the `repr` as soon as there is any hidden field (just like we hide it as soon as there is any private field).

GuillaumeGomez replied:
Seems like there is more to discuss. I'll add it to the next [rustdoc team meeting agenda](https://hackmd.io/ITqOnVbCQaWJzHm5unTjEQ?both).

camelid reviewed (commented):
In discussing this as part of the rustdoc meeting, I realized we probably need to account for whether `#[non_exhaustive]` has been applied to the struct/enum/etc. If it is, then the API isn't stable.

bors commented:
:umbrella: The latest upstream changes (presumably #126788) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dtolnay reviewed (commented):
This appears to be stalled on everyone agreeing on a comprehensive set of rules.

Since I ran into this again today in https://github.com/rust-lang/rust/issues/128364, I'll try to suggest a way to make progress again: Is it possible we can agree to err on the side of **not** showing `repr`? Let's make rustdoc render `repr` in only the most incontrovertible circumstances: everything is pub, nothing is hidden, there is at least 1 field, there is at least one variant, etc. Feel free to add as many other restrictions here as it takes until everyone agrees that we've reached an overestimation of what the actual rules should be.

After that, we can incrementally agree to additional deliberate cases where `repr` should be made part of the documented API of a type. Each of these can be FCP'd with the team if needed.

The current state of erring on the side of rendering `repr` in too many cases that have not been agreed to is unfortunate.

GuillaumeGomez commented:
I think it's good like this. Like @dtolnay mentioned, we can always add new rules later on. Should we start the FCP?

fmease commented:
> FCP

Not quite yet, I still need to update the approach, PR description and PR itself :) I'll do so in a moment. I only rebased.

fmease commented:
Thinking back to past discussions, one reason for being liberal in showing / conservative in hiding `#[repr(...)]` was the fact that there's currently no way to override this heuristic, i.e., force it show when rustdoc hides. Ideas about a `#[doc(repr(...))]` were floating around.

Without it, users would need to declare this information in prose instead.

Thinking aloud, I guess it does make sense as a first step even if it's not greatest (e.g., repr packed and C can be meaningful (as part of the public ABI) even if e.g. later fields are private/hidden as they don't necessarily influence the alignment and thus the offset of earlier public fields (for e.g. `core::mem::offset_of`)).

bors commented:
:umbrella: The latest upstream changes (presumably #129403) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@fmease any updates on this? thanks

fmease commented on src/doc/rustdoc/src/advanced-features.md:
```diff
@@ -89,7 +89,13 @@ https://doc.rust-lang.org/stable/std/?search=%s&go_to_first=true
-## `#[repr(transparent)]`: Documenting the transparent representation
+## `#[repr(...)]`: Documenting the representation of a type
+
+Generally, rustdoc only displays the representation of a given type if none of its variants are
```
I'm very open to phrasing suggestions and ways to make this section and subsection clearer! I feel like it could be much better!

fmease commented:
Only[^2] display the representation[^0] `#[repr(…)]` of a given type if none of its variants[^1] are `#[doc(hidden)]` and if all of its fields are public and not `#[doc(hidden)]` since it's likely not meant to be considered part of the public ABI otherwise.

Moreover, we now also factor in the presence of `#[doc(hidden)]` when checking whether to show `repr(transparent)`.

@rfcbot merge

[^0]: This excludes the representations `Rust` (which we will never display explicitly) and `transparent` (for which we have custom rules, see #115439).
[^1]: Here, *variant* refers to a (general) variant of an ADT which includes not only enum variants but also the single synthetic variant of structs and unions.
[^2]: We're respecting `--document-{private,hidden}-items` of course.

obi1kenobi commented:
I have significant concerns that this may cause unfixable false-positive bugs in cargo-semver-checks.

Doc hidden is not sufficient to indicate non-public-API status — attempting to define it in such a maximalist fashion results in an edge case where major breaking changes are allowed by Rust SemVer rules via a series of only non-major version updates.

Please don't merge this until I've had a chance to review more deeply. Otherwise cargo-semver-checks may be broken in an unfixable fashion.

obi1kenobi commented:
In lieu of being able to list a concern (since I'm not on the review list), I'd like to ask someone on the list to add the concern on my behalf.

fmease committed:
rustdoc: hide `#[repr(...)]` if it isn't part of the public ABI

fmease commented:
I've discussed this with @obi1kenobi in person and we came to the conclusion that we don't want to apply this heuristic / these rules for the JSON output at all. For historical context, we *used* to pass through the `#[repr]` unconditionally to the JSON output until very recently in PR #138018 which regressed that. I'll move the fix for the regression out of this PR.

For additional context, people currently abuse `#[doc(hidden)]` in the ecosystem to hide deprecated functions, hence Predrag's comment about `doc(hidden)` not being a 'reliable' way to denote exclusion from public API/ABI. However, we do want to consider `#[doc(hidden)]` for the HTML output to unblock #116743 because `core`/`std` uses `doc(hidden)` how it's intended to be used (e.g., `c_void` has to 2 hidden variants and a `repr` that's meant to be private).

In a very far & hypothetical future where `core`/`std` use cargo-semver-checks themselves, there would now be a mismatch since CSC won't consider `#[doc(hidden)]` as marking things excluded from the public API/ABI for pragmatic reasons (as mentioned above). However, that would affect every single other item in `core`/`std` that's marked hidden anyway already, so we would need to find a better solution anyway, so it shouldn't be blocking this PR!

@rfcbot resolve positive-bug

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15085015572/job/42406493149) [(plain)](https://github.com/rust-lang/rust/commit/4972dfe27f91fffb4f9ae74387d8d1e45bc98477/checks/42406493149/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.871 Building wheels for collected packages: reuse
#12 2.872   Building wheel for reuse (pyproject.toml): started
#12 3.110   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.111   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#12 3.111   Stored in directory: /tmp/pip-ephem-wheel-cache-stzvng7w/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.114 Successfully built reuse
#12 3.114 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.523 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.524 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.070 Collecting virtualenv
#12 4.115   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.203      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 71.1 MB/s eta 0:00:00
#12 4.266 Collecting filelock<4,>=3.12.2
#12 4.270   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.289 Collecting distlib<1,>=0.3.7
#12 4.293   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.300      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 93.1 MB/s eta 0:00:00
#12 4.336 Collecting platformdirs<5,>=3.9.1
#12 4.341   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.425 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.621 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.621 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      137152 kB
DirectMap2M:     8251392 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 36.279
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/src/librustdoc/clean/types.rs:830:
             let var = adt.variant(rustc_abi::FIRST_VARIANT);
             // `repr(transparent)` is public iff the non-1-ZST field is public or
             // at least one field is public in case all fields are 1-ZST fields.
-            let is_public = is_json || is_visible(var.def_id)
-                && var
-                    .fields
-                    .iter()
-                    .find(|field| {
-                        let ty = field.ty(tcx, ty::GenericArgs::identity_for_item(tcx, field.did));
-                        tcx.layout_of(
-                            ty::TypingEnv::post_analysis(tcx, field.did).as_query_input(ty),
-                        )
-                        .is_ok_and(|layout| !layout.is_1zst())
-                    })
-                    .map_or_else(
-                        || var.fields.is_empty() || var.fields.iter().any(is_field_public),
-                        is_field_public,
-                    );
+            let is_public = is_json
+                || is_visible(var.def_id)
+                    && var
+                        .fields
+                        .iter()
+                        .find(|field| {
+                            let ty =
+                                field.ty(tcx, ty::GenericArgs::identity_for_item(tcx, field.did));
+                            tcx.layout_of(
+                                ty::TypingEnv::post_analysis(tcx, field.did).as_query_input(ty),
+                            )
+                            .is_ok_and(|layout| !layout.is_1zst())
+                        })
+                        .map_or_else(
+                            || var.fields.is_empty() || var.fields.iter().any(is_field_public),
+                            is_field_public,
+                        );
 
             // Since `repr(transparent)` can't have any other reprs or
             // repr modifiers beside it, we can safely return early here.
Diff in /checkout/src/librustdoc/clean/types.rs:863:
             return None;
         }
 
-        let is_public = is_json || adt.variants().iter().all(|variant| {
-            is_visible(variant.def_id) && variant.fields.iter().all(is_field_public)
-        });
+        let is_public = is_json
+            || adt.variants().iter().all(|variant| {
+                is_visible(variant.def_id) && variant.fields.iter().all(is_field_public)
+            });
         if !is_public {
             return None;
         }
fmt: checked 6010 files
Build completed unsuccessfully in 0:01:17
  local time: Sat May 17 12:19:48 UTC 2025
```

</details>

fmease commented:
@rfcbot resolve positive-bugs

GuillaumeGomez commented:
@rfcbot resolve positive-bugs

GuillaumeGomez commented:
@rfcbot resolved positive-bugs

GuillaumeGomez commented:
Bot doesn't seem to want to resolve this concern. ^^'

apiraino commented:
@rfcbot resolve positive-bugs

GuillaumeGomez commented:
@rfcbot resolve positive-bugs in cargo-semver-check

GuillaumeGomez commented:
Fixed. :)

fmease commented on src/librustdoc/clean/types.rs:
```diff
@@ -805,22 +801,40 @@ impl Item {
+        };
+
+        if repr.transparent() {
+            // `repr(transparent)` can only be applied to structs and one-variant enums.
```
```suggestion
            // `repr(transparent)` can only be applied to structs and single-variant enums.
```

fmease commented:
I completely forgot that it was suggested to consider `#[non_exhaustive]`, too. I'll impl that later.

fmease commented:
We now also consider `#[non_exhaustive]`. See the updated PR/FCP description.

@rustbot review

fmease commented:
Checking if there's noticeable query invocation overhead.
@bors2 try @rust-timer queue

obi1kenobi commented:
Just checking: will this affect the JSON output as well, or only HTML?

Also, I wasn't able to find any discussion of why e.g. `repr(i32)` on a `non_exhaustive` enum would be considered non-public. I did see the `c_void` being shown as `repr(u8)` issue, which is certainly unfortunate. But perhaps blanket-banning everything `#[non_exhaustive]` might be an over-correction?

fmease commented:
I'm not married to the non-exhaustive logic, it was suggested here: https://github.com/rust-lang/rust/pull/116882#pullrequestreview-1876246779

fmease commented:
> Just checking: will this affect the JSON output as well, or only HTML?

As previously discussed, we don't and won't apply these heuristics for rustdoc JSON.

rust-timer commented:
Finished benchmarking commit (5e6424bb0d0c85ce18fab29ccab5952ddb21b06b): [comparison URL](https://perf.rust-lang.org/compare.html?start=3bc767e1a215c4bf8f099b32e84edb85780591b1&end=5e6424bb0d0c85ce18fab29ccab5952ddb21b06b&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3bc767e1a215c4bf8f099b32e84edb85780591b1&end=5e6424bb0d0c85ce18fab29ccab5952ddb21b06b&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3bc767e1a215c4bf8f099b32e84edb85780591b1&end=5e6424bb0d0c85ce18fab29ccab5952ddb21b06b&stat=max-rss)
<details>
<summary>Results (primary 5.1%, secondary -2.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.1%  | [5.1%, 5.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-3.0%, -2.3%] | 2     |
| All ❌✅ (primary)                 | 5.1%  | [5.1%, 5.1%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3bc767e1a215c4bf8f099b32e84edb85780591b1&end=5e6424bb0d0c85ce18fab29ccab5952ddb21b06b&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=3bc767e1a215c4bf8f099b32e84edb85780591b1&end=5e6424bb0d0c85ce18fab29ccab5952ddb21b06b&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 757.216s -> 756.086s (-0.15%)
**Artifact size**: 372.11 MiB -> 372.12 MiB (0.00%)

fmease commented:
`#[non_exhaustive]` is no longer part of the ~~equation~~ heuristic. See https://github.com/rust-lang/rust/pull/116882#issuecomment-2978672432 for a rationale.

Please check your boxes https://github.com/rust-lang/rust/pull/116882#issuecomment-2888282009 or voice concerns, repinging @camelid @notriddle.

fmease commented:
r? GuillaumeGomez

fmease committed:
rustdoc: Move HTML-specific attr rendering code into HTML rendering mod

These functions used to be shared with the JSON backend but nowadays they
aren't.

fmease committed:
rustdoc: Slightly clean up attr rendering

fmease committed:
rustdoc: Fully escape link section and export name

Escape "special characters" (e.g., double quotes `"` and line breaks `\n`).
Escape HTML.

Lastly, add regression tests and clean up existing tests.

GuillaumeGomez commented:
Time to merge, thanks everyone!

@bors r=rustdoc rollup
